core/domain/exp_jobs_one_off.py

```diff
@@ -728,3 +728,183 @@ def map(item):
+        if key.startswith('SUCCESS'):
+            yield (key, len(values))
+        else:
+            yield (key, values)
+
+
+class ExpSnapshotsMigrationJob(jobs.BaseMapReduceOneOffJobManager):
+    """A reusable one-time job that may be used to migrate exploration schema
+    versions. This job will load all snapshots of all existing explorations
+    from the datastore and immediately store them back into the datastore.
+    The loading process of an exploration in exp_services automatically
+    performs schema updating. This job persists that conversion work, keeping
+    explorations up-to-date and improving the load time of new explorations.
+
+    NOTE TO DEVELOPERS: Make sure to run ExpSnapshotsMigrationAuditJob before
+    running this job.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationSnapshotContentModel]
+
+    @classmethod
+    def enqueue(cls, job_id, additional_job_params=None):
+        super(ExpSnapshotsMigrationJob, cls).enqueue(
+            job_id, shard_count=64)
+
+    @staticmethod
+    def map(item):
+        exp_id = item.get_unversioned_instance_id()
+
+        latest_exploration = exp_fetchers.get_exploration_by_id(
+            exp_id, strict=False)
+        if latest_exploration is None:
+            yield ('INFO - Exploration does not exist', item.id)
+            return
+
+        exploration_model = exp_models.ExplorationModel.get(exp_id)
+        if (exploration_model.states_schema_version !=
+                feconf.CURRENT_STATE_SCHEMA_VERSION):
+            yield (
+                'FAILURE - Exploration is not at latest schema version', exp_id)
+            return
+
+        try:
+            latest_exploration.validate()
+        except Exception as e:
+            yield (
+                'INFO - Exploration %s failed non-strict validation' % item.id,
+                e)
+
+        # Some old explorations do not have a states schema version.
+        if 'states_schema_version' not in item.content:
+            item.content['states_schema_version'] = 0
+
+        # If the snapshot being stored in the datastore does not have the most
+        # up-to-date states schema version, then update it.
+        target_state_schema_version = feconf.CURRENT_STATE_SCHEMA_VERSION
+        current_state_schema_version = item.content['states_schema_version']
+        if current_state_schema_version == target_state_schema_version:
+            yield (
+                'SUCCESS - Snapshot is already at latest schema version',
+                item.id)
+            return
+
+        versioned_exploration_states = {
+            'states_schema_version': current_state_schema_version,
+            'states': item.content['states']
+        }
+        while current_state_schema_version < target_state_schema_version:
+            assert (
+                versioned_exploration_states['states_schema_version'] ==
+                current_state_schema_version)
```

Comment:
This would crash the job with an exception if it failed, shouldn't we yield an error and return early instead?

Reply:
Nah. It won't fail; the logic ensures that and I've already done a test run. This is just there out of "abundance of caution".

Reply:
Ack'd that it's an abundance of caution, but then it seems overkill to use `assert` since it would crash the job. What's wrong with a yield followed by return?

Reply:
Oh sorry, I forgot to also say that I'm deleting it :P (now done!)

Reply:
Cool, thanks!